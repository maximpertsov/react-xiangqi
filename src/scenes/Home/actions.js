import * as client from 'services/client';

export const fetchGames = ({ username }) => async dispatch => {
  if (username === null) return;

  const {
    data: { games },
  } = await client.getGameList({ username });
  dispatch({ type: 'set_games', games });
};

export const setForm = form => dispatch => {
  const { username, password, error } = form;
  if (username !== undefined) {
    dispatch({ type: 'set_login_username', username });
  }
  if (password !== undefined) {
    dispatch({ type: 'set_login_password', password });
  }
  if (error !== undefined) {
    dispatch({ type: 'set_login_error', error });
  }
};

export default {};
