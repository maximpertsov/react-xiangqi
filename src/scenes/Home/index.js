import React, { useCallback, useEffect, useState } from 'react';
import { Button, Container, Header, Segment } from 'semantic-ui-react';

import Game from 'scenes/Game';
import * as client from 'services/client';

import GameList from './components/GameList';
import LoginForm from './components/LoginForm';

// TODO: absolute import?
import { AutoMove } from '../../constants';

const LOCAL = 'local';
const PLAYER_VS_CPU = 'player_vs_cpu';
const CPU_VS_CPU = 'cpu_vs_cpu';

const Home = () => {
  const [username, setUsername] = useState(undefined);
  const [gameSlug, setGameSlug] = useState(undefined);
  const [games, setGames] = useState([]);

  const fetchGames = useCallback(
    async() => {
      const response = await client.getGameList(username);
      setGames(response.data.games);
    },
    [username],
  );

  useEffect(
    () => {
      if (username === undefined) return;
      fetchGames();
    },
    [fetchGames, username],
  );

  const renderGameList = () => (
    username === undefined ||
    <GameList setGameSlug={setGameSlug} games={games} />
  );

  const renderMenu = () => (
    <Container textAlign="center">
      <Segment.Group compact>
        <Segment>
          <Header size="large">Play online</Header>
          <LoginForm
            username={username}
            setUsername={setUsername}
          />
          { renderGameList() }
        </Segment>
        <Segment>
          <Header size="large">Other modes</Header>
          <Button onClick={() => { setGameSlug(LOCAL); }}>
            Solo play
          </Button>
          <Button onClick={() => { setGameSlug(PLAYER_VS_CPU); }}>
            vs CPU
          </Button>
          <Button onClick={() => { setGameSlug(CPU_VS_CPU); }}>
            CPU vs CPU
          </Button>
        </Segment>
      </Segment.Group>
    </Container>
  );

  switch (gameSlug) {
    case undefined:
      return renderMenu();
    case LOCAL:
      return <Game />;
    case PLAYER_VS_CPU:
      return <Game autoMove={AutoMove.BLACK} />;
    case CPU_VS_CPU:
      return <Game autoMove={AutoMove.BOTH} />;
    default:
      return (
        <Game
          gameSlug={gameSlug}
          username={username}
        />
      );
  }
};

export default Home;
