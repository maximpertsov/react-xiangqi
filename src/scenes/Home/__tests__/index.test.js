import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';

import { applyMiddleware, createStore, compose } from 'redux';
import rootReducer from 'reducers';

import Home from 'scenes/Home';

const initialState = {
  gameSlug: null,
  showGame: false,
};

const getStore = (overrides = {}) =>
  compose(applyMiddleware(thunk))(createStore)(rootReducer, {
    ...initialState,
    ...overrides,
  });

const getComponent = store => (
  <Provider store={store}>
    <Home />
  </Provider>
);

describe('Home', () => {
  test('click solo play without crashing', () => {
    const wrapper = mount(getComponent(getStore()));
    expect(wrapper.render()).toMatchSnapshot();

    wrapper
      .findWhere(node => node.is('.button') && node.text() === 'Solo play')
      .simulate('click');
    jest.runOnlyPendingTimers();
    wrapper.update();

    // TODO: snapshot shows blurring because state is missing
    expect(wrapper.render()).toMatchSnapshot();
    wrapper.unmount();
  });
});
