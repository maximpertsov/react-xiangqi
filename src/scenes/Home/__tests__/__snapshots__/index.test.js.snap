// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home click solo play without crashing 1`] = `
<div
  class="ui center aligned container"
>
  <div
    class="ui compact segments"
  >
    <div
      class="ui segment"
    >
      <div
        class="ui large header"
      >
        Play online
      </div>
      <div
        class="LoginForm"
      >
        <form
          class="ui form"
        >
          <div
            class="field"
          >
            <label>
              Username
            </label>
            <div
              class="ui left icon input"
            >
              <input
                name="username"
                placeholder="Username"
                type="text"
                value=""
              />
              <i
                aria-hidden="true"
                class="user icon"
              />
            </div>
          </div>
          <div
            class="field"
          >
            <label>
              Password
            </label>
            <div
              class="ui left icon input"
            >
              <input
                name="password"
                type="password"
                value=""
              />
              <i
                aria-hidden="true"
                class="lock icon"
              />
            </div>
          </div>
          <button
            class="ui primary button"
          >
            Login
          </button>
        </form>
        <div />
      </div>
    </div>
    <div
      class="ui segment"
    >
      <div
        class="ui medium header"
      >
        Create game
      </div>
      <div
        class="NewGameMenu css-1hstsbv"
      >
        <div
          class="ui fluid buttons"
        >
          <button
            class="ui icon button"
          >
            <i
              aria-hidden="true"
              class="red plus square outline fitted icon"
            />
          </button>
          <button
            class="ui icon button"
          >
            <i
              aria-hidden="true"
              class="black plus square outline fitted icon"
            />
          </button>
          <button
            class="ui icon button"
          >
            <i
              aria-hidden="true"
              class="question fitted icon"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="ui segment"
    >
      <div
        class="Lobby css-8bzhik"
      >
        <div
          class="ui medium header"
        >
          Lobby
        </div>
        <div
          class="css-1vrdtb6"
        />
      </div>
    </div>
    <div
      class="ui segment"
    >
      <div
        class="ui large header"
      >
        Other modes
      </div>
      <button
        class="ui button"
      >
        Solo play
      </button>
      <button
        class="ui button"
      >
        vs CPU
      </button>
      <button
        class="ui button"
      >
        CPU vs CPU
      </button>
    </div>
  </div>
</div>
`;

exports[`Home click solo play without crashing 2`] = `
<div>
  <div
    class="ui basic segment blurring dimmed dimmable"
  />
</div>
`;
